File: C#/ShellSort_1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_
{
    internal class ShellSort_1 : MethodSort
    {
        /// <summary>
        /// シェルソート1（改良挿入法）
        /// </summary>
        public override void Sort()
        {
            Random rnd = new Random();
            int[] data = new int[100];

            for(int i = 0; i < data.Length; i++)
            {
                data[i] = rnd.Next(100);
            }

            int gap = data.Length / 2;

            while(gap > 0)
            {
                // ギャップとびの部分数列のソート
                for(int k = 0; k < gap; k++)
                {
                    for(int i = k + gap; i < data.Length; i += gap)
                    {
                        for(int j = i - gap; j > k - 1; j -= gap)
                        {
                            if (data[j] > data[j + gap])
                            {
                                int temp = data[j];
                                data[j] = data[j + gap];
                                data[j + gap] = temp;
                            }
                            else
                            {
                                break;
                            }
                        }
                    }
                }

                gap /= 2;
            }

            for(int i = 0; i < data.Length; i++)
            {
                Console.Write(data[i] + " ");
            }

            Console.WriteLine();
        }
    }
}

File: Java/Sort/src/logic/ShellSort_1.java
package logic;

import java.util.Random;

public class ShellSort_1 extends LogicSort {
	// シェルソート（改良挿入法）1
	public void Sort() {
		int[] data = new int[100];
		int dataLength = data.length;

		Random rnd = new Random();
		for (int i = 0; i < dataLength; i++) {
			data[i] = rnd.nextInt(100);
		}

		int gap = dataLength / 2;
		int replaceCount = 0;
		
		while (gap > 0) {
			for (int k = 0; k < gap; k++) {
				for (int i = k + gap; i < dataLength; i += gap) {
					for (int j = i - gap; j > -1; j -= gap) {
						if (data[j] > data[j + gap]) {
							replaceCount++;
							
							int temp = data[j];
							data[j] = data[j + gap];
							data[j + gap] = temp;
						} else {
							break;
						}
					}
				}
			}

			gap /= 2;
		}

		for (int i = 0; i < dataLength; i++) {
			System.out.print(data[i] + " ");
		}

		System.out.println();
		
		System.out.println("入替回数:" + replaceCount);
	}
}
